
..
 pipeline
  what this is

 quick start
 

Variant Calling and core SNP diagnostics Pipeline
=================================================

.. toctree::
   :hidden:

   pipeline
   quick-start
   command-line-options
   results
   variant-diagnostics
   bonus-ducks
   config
   logs
   


`Canu <http://github.com/marbl/canu>`_ is a fork of the Celera Assembler designed for high-noise single-molecule sequencing (such as
the PacBio RSII or Oxford Nanopore MinION).

This `pipeline <https://github.com/alipirani88/variant_calling_pipeline>`_ calls variants on PE/SE reads provided in a directory and generates core SNP/consensus fasta files that can be used to to build a phylogeny or as an input for Gubbins/Beast analysis.


Install
=========
Ignore this if you are in snitkin lab. The dependencies are already installed in lab bin_group folder:
::
    /nfs/esnitkin/bin_group/variant_calling_bin/

Use the python version installed in:
::
    /nfs/esnitkin/bin_group/anaconda2/bin/python

Links
=========

*  :ref:`Pipeline		<pipeline>` - Describes different steps involved in this pipeline.
*  :ref:`Quick Start		<quick-start>` - A quick-start guide to run pipeline on the reads.
*  :ref:`Command Line Options	<command-line-options>`   - command line options to customize the run.
*  :ref:`Results		<results>`   - Results generated by the pipeline
*  :ref:`Variant Diagnostics	<variant-diagnostics>` - Generating plots to diagnose variant calls.
*  :ref:`Bonus Ducks		<bonus-ducks>` - tricks and shortcuts.
*  :ref:`Config			<config>` - Customization and description of config file
*  :ref:`Logs              	<logs>` - log description

